### PR TITLE
Support jwt

### PR DESCRIPTION
**Pull request by RileyYe**
## What I did

Support JWT and block unauthorized users from creating news.

## Checklist

Hello! Welcome to send a PR to us! Here is a simple PR checklist:

- [ ] Format your code, by `yarn format`. A good format is helpful for
      everyone.
- [ ] (NOT NOW) Make sure tests are passed.
- [ ] If you need to update the document, then update. The documents are here:
  - `README.md` in the root folder of the project, about what the project is and
    why the project exists.
  - `README.md` in the package `back`, `front` and `theme-shortcut`(at
    the path `front/packages/theme-shortcut`).
  - In folder `docs`, there are the other documents.

### CODE REVIEW

File: back/src/graphql/type/user.js [hunk 1]
```diff
@@ -36,7 +37,7 @@ const getUserNumber = async () => {
   const [rows, _fields] = await conn.execute(SQL_USER_NUMBER);
   // It must be only one line.
   const row = rows[0];
-  return row.user_number;
+  return row.userNumber;
```
Comment (PeterlitsZo) [inline back/src/graphql/type/user.js:40]: It should be changed on #20.

File: back/src/graphql/type/user.js [hunk 2]
```diff
@@ -146,7 +147,20 @@ export const signIn = {
...
+        isAdmin: res.isAdmin,
+      },
+      salt,
+      {
+        expiresIn: '15d',
+        algorithm: 'HS256',
+      },
+    );
+    console.log(token);
+    context.res.cookie('jwt', token, { httpOnly: true });
```
Comment (PeterlitsZo) [inline back/src/graphql/type/user.js:162]: Great!!!

**Comment by RileyYe** [comment]
err maybe next night I will fix them，may cost me a while





------------------ Original ------------------
From: peterlits zo ***@***.***&gt;
Date: Thu,Mar 24,2022 10:06 PM
To: ACMHomepage/ACMHomepage ***@***.***&gt;
Cc: RileyYe ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [ACMHomepage/ACMHomepage] Support jwt (PR #26)





 
@PeterlitsZo requested changes on this pull request.
 
 
In back/src/main.js:
 &gt; @@ -25,7 +31,7 @@ GraphQLServer.use(    (req, res) =&gt; {      return graphqlHTTP({        schema: GraphQLSchema, -      graphiql: GRAPHIQL, +      graphiql: true,  
In my opinion, we should only open GraphiQL if and only if we are in the development env.
 
 
In back/src/graphql/type/news.js:
 &gt; -  async resolve(_parentVal, args) { +  async resolve(_parentVal, args, context) { +    const verified = await jwt.verify( +      context.req.cookies.jwt, +      salt, +      (err, decoded) =&gt; { +        if (err) { +          console.log('Error!'); +          return false; +        } else if (decoded.isAdmin) { +          return true; +        } +      }, +    ); +    if (!verified) { +      return {};  
I think to throw an Error we be better.
 
see the error in GraphQL.
 
 
In back/src/graphql/type/user.js:
 &gt; @@ -36,7 +37,7 @@ const getUserNumber = async () =&gt; {    const [rows, _fields] = await conn.execute(SQL_USER_NUMBER);    // It must be only one line.    const row = rows[0]; -  return row.user_number; +  return row.userNumber;  
It should be changed on #20.
 
 
In back/src/graphql/type/user.js:
 &gt; @@ -105,12 +106,12 @@ export const register = {        VALUES (?, ?, ?, ?)      `;      // TODO: Use bcrypt rather than plain text. -    const userNumber = getUserNumber(); +    const userNumber = await getUserNumber();  
#20.
 
 
In back/src/graphql/type/user.js:
 &gt; @@ -146,7 +147,20 @@ export const signIn = {        throw new Error('The password is not right.');      }      const res = (await getUserById(id))[0]; -    context.res.cookie('jwt', `${res.email}`); +    const token = jwt.sign( +      { +        id: res.id, +        isAdmin: res.isAdmin, +      }, +      salt, +      { +        expiresIn: '15d', +        algorithm: 'HS256', +      }, +    ); +    console.log(token); +    context.res.cookie('jwt', token, { httpOnly: true });  
Great!!!

**Comment by PeterlitsZo** [comment]
> By the way, where is the PR template? Do you remove it?

I edited your PR message.

**Comment by PeterlitsZo** [comment]
Please rebase your branch to `master`. I just merged #20.

**Comment by RileyYe** [comment]
> Please rebase your branch to `master`. I just merged #20.

I got it.

**Comment by RileyYe** [comment]
> Please rebase your branch to `master`. I just merged #20.

Shall I `git checkout support-jwt` then `git rebase master ` ?

**Comment by PeterlitsZo** [comment]
I am not sure... I will run `git switch support-jwt && git rebase -i master`

**Review by PeterlitsZo — APPROVED** [review]
Great work.

File: back/src/graphql/type/news.js [hunk 3]
```diff
@@ -7,7 +7,8 @@ import {
   GraphQLID,
   GraphQLNonNull,
 } from 'graphql';
-
+import { salt } from '../../main.js';
```
Comment (PeterlitsZo) [inline back/src/graphql/type/news.js:10]: Where is the salt in the file `main.js`? Do you update it?

**Review by PeterlitsZo — COMMENTED** [review]
The second commit message: Please do not wrap the commit title.

**Pull request merged**